README
Test::File::ShareDir

"Test::File::ShareDir" is some low level plumbing to enable a
distribution to perform tests while consuming its own "share"
directories in a manner similar to how they will be once installed.

This allows "File::ShareDir" to see the *latest* version of content
instead of simply whatever is installed on whichever target system you
happen to be testing on.

Note: This module only has support for creating 'new' style share dirs
and are NOT compatible with old File::ShareDirs.

For this reason, unless you have File::ShareDir 1.00 or later installed,
this module will not be usable by you.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Test-File-ShareDir

1.001003 2017-03-01T11:38:51Z e2e623c

1.001002 2017-03-01T11:17:51Z 44658a2
 [Dependencies::Stats]
 - Dependencies changed since 1.001001, see misc/*.deps* for details
 - develop: +3 ↑4 (suggests: ↑2)

 [Testfix]
 - Update tooling to aovid .-in-@INC issues
 - Closes https://github.com/kentnl/Test-File-ShareDir/issues/3

1.001001 2015-03-01T05:44:37Z a89c745
 [TestFixes]
 - Try to guard against accidentally failing if a test distdir does in fact exist in the "Expect not to exist" test.

1.001000 2015-02-28T11:09:38Z d703dd2
 [Dependencies::Stats]
 - Dependencies changed since 1.000005, see misc/*.deps* for details
 - configure: +1 (recommends: -1, suggests: +1)
 - develop: +10 ↑3 -2 (suggests: +1 ↑2)
 - runtime: +3 (suggests: +1)
 - test: +1 (recommends: -2, suggests: +2)

 [Deprecations]
 - OO interfaces have a ->add_to_inc method which is being replaced with the more generic ->register, which wil retain
   meaning when the interface no longer actually uses @INC

 [Features]
 - OO interfaces now all have a ->register and a ->clear method.
 - Now supports exporting with_dist_dir and with_module_dir functions from Test::File::ShareDir

1.000005 2014-08-15T23:41:58Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - tests augmented.
 - whitespace greatly reduced.

 [Dependencies::Stats]
 - Dependencies changed since 1.000004, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 (suggests: ↑1)
 - test: -4 (recommends: ↑1 -2)

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - update Test::ReportPrereqs

1.000004 2014-07-22T19:29:27Z
 [Doc Bugfix]
 - Previous releases changelog was misleading.
 - We're requiring a newer Path::Tiny to avoid a problem. ( As opposed to the impression that a newer Path::Tiny caused
   the problem, grammar-- )

1.000003 2014-07-22T19:02:06Z
 [Bugfix]
 - Attempt to avoid tempdir(CLEANUP => 1) being an error by requiring newer Path::Tiny

 [Dependencies::Stats]
 - Dependencies changed since 1.000002, see misc/*.deps* for details
 - runtime: ↑1

1.000002 2014-07-18T16:33:36Z
 [00 Trivial]
 - Minor packaging cleanups shaving ~7k

 [Dependencies::Stats]
 - Dependencies changed since 1.000001, see misc/*.deps* for details
 - develop: (suggests: ↑2)
 - test: +2 ↑1 (recommends: +1)

1.000001 2014-06-07T15:14:38Z
 [00 Trivial]
 - Mostly internal aesthetic changes / perlcritic things.
 - tempdir behaviours used via Path::Tiny interface instead of directly.
 - Silly runtime function proxies mostly deleted.
 - packaging tweaked minorly

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - configure: (recommends: ↑1)
 - develop: +3 -3 (suggests: +2 -1)
 - runtime: -1
 - test: (recommends: ↑2)

1.000000 2014-01-30T15:23:48Z
 [00 Tooling and Version replacement]
 - Switched to EUMM to accelerate clean-install time.
 - Build-time requirements are now softened
 - Readvertise higher requirements as recommendations
 - version schema moved from x.y.z to x.YYYzzz and is the primary reason for incrementing the major

 [Dependencies::Stats]
 - Dependencies changed since 0.4.1, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +7 ↑1 -9
 - test: +4 ↓1 (recommends: +4)

0.4.1 2013-11-21T06:51:24Z
 [00 Trivial Doc Fix]
 - Escaping of code containing > was done wrong.

0.4.0 2013-11-21T06:38:44Z
 [00 Major Changes]
 - New Modules/Interfaces to make life easier for people who only need one of (Dist|Module) shares.

 [Dependencies::Noteworthy]
 - use Class::Tiny
 - use new MB and TestMore

 [Dependencies::Stats]
 - Dependencies changed since 0.3.3, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +45 -1 (recommends: -1, suggests: +1 -1)
 - runtime: +1
 - test: +1 ↑1 -1

 [Documentation]
 - Encoding specified
 - MetaPOD annotations added

 [Packaging]
 - use new MB syntax for fallback

 [Tests]
 - use Test::Compile::PerFile
 - Regenerate ReportVersions::Tiny to not report develop deps.
 - Regenerate CPAN::Changes test
 - Regenerate Kwalitee test

0.3.3 2013-04-03T11:32:08Z
 [!Noteable Changes]
 - Replace Path::Class with Path::Tiny

 [Dependencies::Noteworthy]
 - use Path::Tiny
 - stop using Path::Class::Dir

 [Dependencies::Stats]
 - Dependencies changed since 0.3.2, see misc/*.deps* for details
 - runtime: +1 -1

0.3.2 2013-04-03T07:03:58Z
 - Maintenance release for MB 0.4004

 [Dependencies::Noteworthy]
 - 5.006 is now minimum
 - upgrade MB and TM

 [Dependencies::Stats]
 - Dependencies changed since 0.3.1, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +5 (recommends: ↑1, suggests: ↑1)
 - runtime: +1
 - test: ↑1

 [Documentation]
 - Reindent License
 - Update Copyright Year
 - Update Copyright address
 - Add README.mkdn

 [Meta]
 - Bug tracker moved to github issues

 [Packaging]
 - Build.PL updated for 0.4004 / test_requires

 [Tracking]
 - This source version of 0.3.1 lost and reconstructed

0.3.1 2011-10-28T20:58:28Z
 [Bugs]
 - Absolute paths were treated incorrectly as relative ones. Now using Path::Class properly which should fix this
   problem. Thanks to RJBS for diagnosing and reporting this problem.

 [Dependencies::Noteworthy]
 - test removed: English

 [Dependencies::Stats]
 - Dependencies changed since 0.3.0, see misc/*.deps* for details
 - runtime: +2
 - test: -1

 [Packaging]
 - Updated License ( Indentation, Address )
 - Extra tests moved to xt/

0.3.0 2011-05-05T03:29:45Z
 [Dependencies::Noteworthy]
 - For testing, Now requires Cwd as provided by File-Spec / PathTools.

 [Dependencies::Stats]
 - Dependencies changed since 0.2.0, see misc/*.deps* for details
 - test: +1

 [Documentation]
 - Improved primary documentation to explain import flags

 [Features]
 - In the absense of passing the 'root' parameter, it will default to the current work directory. Considering that most

lib/Test/File/ShareDir.pm
use 5.006;    # pragmas
use strict;
use warnings;

package Test::File::ShareDir;

our $VERSION = '1.001003';

# ABSTRACT: Create a Fake ShareDir for your modules for testing.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY












use File::ShareDir 1.00 qw();
use Exporter qw();
use Test::File::ShareDir::Utils qw( extract_dashes );
use Carp qw( croak );
use parent qw( Exporter );

our @EXPORT_OK = qw( with_dist_dir with_module_dir );

sub import {
  my ( $package, @args ) = @_;

  my ( @imports, %params );

  # ->import( {  }, qw( imports ) )
  if ( 'HASH' eq ref $args[0] ) {
    %params  = %{ shift @args };
    @imports = @args;
  }
  else {
    # ->import( -arg  => value, -arg => value, @imports );
    while (@args) {
      if ( $args[0] =~ /\A-(.*)\z/msx ) {
        $params{ $args[0] } = $args[1];
        splice @args, 0, 2, ();
        next;
      }
      push @imports, shift @args;
    }
  }

  if ( keys %params ) {
    require Test::File::ShareDir::TempDirObject;

    my $tempdir_object = Test::File::ShareDir::TempDirObject->new( \%params );

    for my $module ( $tempdir_object->_module_names ) {
      $tempdir_object->_install_module($module);
    }

    for my $dist ( $tempdir_object->_dist_names ) {
      $tempdir_object->_install_dist($dist);
    }

    unshift @INC, $tempdir_object->_tempdir->stringify;

  }
  if (@imports) {
    $package->export_to_level( 1, undef, @imports );
  }
  return;

}

# This code is just to make sure any guard objects
# are not lexically visible to the sub they contain creating a self reference.
sub _mk_clearer {
  my ($clearee) = @_;
  return sub { $clearee->clear };
}


























sub with_dist_dir {
  my ( $config, $code ) = @_;
  if ( 'CODE' ne ( ref $code || q{} ) ) {
    croak( 'CodeRef expected at end of with_dist_dir(), ' . ( ref $code || qq{scalar="$code"} ) . ' found' );
  }
  require Test::File::ShareDir::Object::Dist;
  require Scope::Guard;
  my $dist_object = Test::File::ShareDir::Object::Dist->new( extract_dashes( 'dists', $config ) );
  $dist_object->install_all_dists();
  $dist_object->register();
  my $guard = Scope::Guard->new( _mk_clearer($dist_object) );    ## no critic (Variables::ProhibitUnusedVarsStricter)
  return $code->();
}


























sub with_module_dir {
  my ( $config, $code ) = @_;
  if ( 'CODE' ne ( ref $code || q{} ) ) {
    croak( 'CodeRef expected at end of with_module_dir(), ' . ( ref $code || qq{scalar="$code"} ) . ' found' );
  }

  require Test::File::ShareDir::Object::Module;
  require Scope::Guard;

  my $module_object = Test::File::ShareDir::Object::Module->new( extract_dashes( 'modules', $config ) );

  $module_object->install_all_modules();
  $module_object->register();
  my $guard = Scope::Guard->new( _mk_clearer($module_object) );    ## no critic (Variables::ProhibitUnusedVarsStricter)

  return $code->();
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::File::ShareDir - Create a Fake ShareDir for your modules for testing.

=head1 VERSION

version 1.001003

=head1 SYNOPSIS

    use Test::More;

    # use FindBin; optional

    use Test::File::ShareDir
        # -root => "$FindBin::Bin/../" # optional,
        -share => {
            -module => { 'My::Module' => 'share/MyModule' }
            -dist   => { 'My-Dist'    => 'share/somefolder' }
        };

    use My::Module;

    use File::ShareDir qw( module_dir dist_dir );

    module_dir( 'My::Module' ) # dir with files from $dist/share/MyModule

    dist_dir( 'My-Dist' ) # dir with files from $dist/share/somefolder

=head1 DESCRIPTION

C<Test::File::ShareDir> is some low level plumbing to enable a distribution to perform tests while consuming its own C<share>
directories in a manner similar to how they will be once installed.

This allows C<File::ShareDir> to see the I<latest> version of content instead of simply whatever is installed on whichever target
system you happen to be testing on.

B<Note:> This module only has support for creating 'new' style share dirs and are NOT compatible with old File::ShareDirs.

For this reason, unless you have File::ShareDir 1.00 or later installed, this module will not be usable by you.

=head1 SIMPLE INTERFACE

Starting with version C<0.4.0>, there are a few extra interfaces you can use.

These will probably be more useful, and easier to grok, because they don't have a layer of
indirection in order to simultaneously support both C<Module> and C<Dist> C<ShareDir>'s.

=head2 Simple Exporter Interfaces

=head3 C<Test::File::ShareDir::Dist>

L<< C<Test::File::ShareDir::Dist>|Test::File::ShareDir::Dist >> provides a simple export interface
for making C<TempDir> C<ShareDir>'s from a given path:

    use Test::File::ShareDir::Dist { "Dist-Name" => "share/" };

This will automatically create a C<ShareDir> for C<Dist-Name> in a C<TempDir> based on the contents of C<CWD/share/>

See L<< C<Test::File::ShareDir::Dist>|Test::File::ShareDir::Dist >> for details.

=head3 C<Test::File::ShareDir::Module>

L<< C<Test::File::ShareDir::Module>|Test::File::ShareDir::Module >> provides a simple export interface
for making C<TempDir> C<ShareDir>'s from a given path:

    use Test::File::ShareDir::Module { "Module::Name" => "share/" };

This will automatically create a C<ShareDir> for C<Module::Name> in a C<TempDir> based on the contents of C<CWD/share/>

See L<< C<Test::File::ShareDir::Module>|Test::File::ShareDir::Module >> for details.

=head2 Simple Object Oriented Interfaces

=head3 C<Test::File::ShareDir::Object::Dist>

L<< C<Test::File::ShareDir::Object::Dist>|Test::File::ShareDir::Object::Dist >> provides a simple object oriented interface for
making C<TempDir> C<ShareDir>'s from a given path:

    use Test::File::ShareDir::Object::Dist;

    my $obj = Test::File::ShareDir::Object::Dist->new( dists => { "Dist-Name" => "share/" } );
    $obj->install_all_dists;
    $obj->register;

This will automatically create a C<ShareDir> for C<Dist-Name> in a C<TempDir> based on the contents of C<CWD/share/>

See L<< C<Test::File::ShareDir::Object::Dist>|Test::File::ShareDir::Object::Dist >> for details.

=head3 C<Test::File::ShareDir::Object::Module>

L<< C<Test::File::ShareDir::Object::Module>|Test::File::ShareDir::Object::Module >> provides a simple object oriented interface
for making C<TempDir> C<ShareDir>'s from a given path:

    use Test::File::ShareDir::Object::Module;

    my $obj = Test::File::ShareDir::Object::Module->new( modules => { "Module::Name" => "share/" } );
    $obj->install_all_modules;
    $obj->register;

This will automatically create a C<ShareDir> for C<Module::Name> in a C<TempDir> based on the contents of C<CWD/share/>

See L<< C<Test::File::ShareDir::Object::Module>|Test::File::ShareDir::Object::Module >> for details.

=head1 SCOPE LIMITED UTILITIES

C<Test::File::ShareDir> provides a few utility functions to aide in temporarily adjusting C<ShareDir> behavior.

    use Test::File::ShareDir qw( with_dist_dir with_module_dir );

    with_dist_dir({ 'Dist-Name' => 'Some/Path' }, sub {
      # dist_dir() now behaves differently here
    });
    with_module_dir({ 'Module::Name' => 'Some/Path' }, sub {
      # module_dir() now behaves differently here
    });

See L<< C<EXPORTABLE FUNCTIONS>|/EXPORTABLE FUNCTIONS >> for details.

=head1 IMPORTING

Since C<1.001000>, there are 2 ways of passing arguments to C<import>

  use Foo { -root => ... options }, qw( functions to import );
  use Foo -optname => option, -optname => option, qw( functions to import );

Both should work, but the former might be less prone to accidental issues.

=head2 IMPORT OPTIONS

=head3 -root

This parameter is the prefix the other paths are relative to.

If this parameter is not specified, it defaults to the Current Working Directory ( C<CWD> ).

In versions prior to C<0.3.0>, this value was mandatory.

The rationale behind using C<CWD> as the default value is as follows.

=over 4

=item * Most users of this module are likely to be using it to test distributions

=item * Most users of this module will be using it in C<$project/t/> to load files from C<$project/share/>

=item * Most C<CPAN> tools run tests with C<CWD> = $project

=back

Therefor, defaulting to C<CWD> is a reasonably sane default for most people, but where it is not it can
still be overridden.

  -root => "$FindBin::Bin/../" # resolves to project root from t/ regardless of Cwd.

=head3 -share

This parameter is mandatory, and contains a C<hashref> containing the data that explains what directories you want shared.

  -share =>  { ..... }

=head4 -module

C<-module> contains a C<hashref> mapping Module names to path names for module_dir style share dirs.

  -share => {
    -module => { 'My::Module' => 'share/mymodule/', }
  }

  ...

  module_dir('My::Module')

Notedly, it is a C<hashref>, which means there is a limitation of one share dir per module. This is simply because having more
than one share dir per module makes no sense at all.

=head4 -dist

C<-dist> contains a C<hashref> mapping Distribution names to path names for dist_dir style share dirs. The same limitation
applied to C<-module> applies here.

  -share => {
    -dist => { 'My-Dist' => 'share/mydist' }
  }
  ...
  dist_dir('My-Dist')

=head1 EXPORTABLE FUNCTIONS

=head2 with_dist_dir

Sets up a C<ShareDir> environment with limited context.

  # with_dist_dir(\%config, \&sub);
  with_dist_dir( { 'Dist-Name' => 'share/' } => sub {

      # File::ShareDir resolves to a copy of share/ in this context.

  } );

C<%config> can contain anything L<< C<Test::File::ShareDir::Dist>|Test::File::ShareDir::Dist >> accepts.

=over 4

=item C<-root>: Defaults to C<$CWD>

=item C<I<$distName>>: Declare C<$distName>'s C<ShareDir>.

=back

I<Since 1.001000>

=head2 with_module_dir

Sets up a C<ShareDir> environment with limited context.

  # with_module_dir(\%config, \&sub);
  with_module_dir( { 'Module::Name' => 'share/' } => sub {

      # File::ShareDir resolves to a copy of share/ in this context.

  } );

C<%config> can contain anything L<< C<Test::File::ShareDir::Module>|Test::File::ShareDir::Module >> accepts.

=over 4

=item C<-root>: Defaults to C<$CWD>

=item C<I<$moduleName>>: Declare C<$moduleName>'s C<ShareDir>.

=back

I<Since 1.001000>

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"Test::File::ShareDir",
    "interface":"exporter"
}


=end MetaPOD::JSON

=head1 THANKS

Thanks to the C<#distzilla> crew for ideas,suggestions, code review and debugging, even though not all of it made it into releases.

=for stopwords DOLMEN ETHER HAARG RJBS

=over 4

=item * L<DOLMEN|cpan:///author/dolmen>

=item * L<ETHER|cpan:///author/ether>

=item * L<HAARG|cpan:///author/haarg>

=item * L<RJBS|cpan:///author/rjbs>

=back

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Test/File/ShareDir/Dist.pm
use 5.006;    # pragmas
use strict;
use warnings;

package Test::File::ShareDir::Dist;

our $VERSION = '1.001003';

# ABSTRACT: Simplified dist oriented ShareDir tester

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use File::ShareDir 1.00 qw();
use Test::File::ShareDir::Utils qw( extract_dashes );












sub import {
  my ( undef, $arg ) = @_;

  if ( not ref $arg or 'HASH' ne ref $arg ) {
    require Carp;
    return Carp::croak q[Must pass a hashref];
  }

  my %input_config = %{$arg};

  require Test::File::ShareDir::Object::Dist;

  my $dist_object = Test::File::ShareDir::Object::Dist->new(extract_dashes('dists', \%input_config ));
  $dist_object->install_all_dists();
  $dist_object->register();
  return 1;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::File::ShareDir::Dist - Simplified dist oriented ShareDir tester

=head1 VERSION

version 1.001003

=head1 SYNOPSIS

    use Test::File::ShareDir::Dist {
        '-root' => 'some/root/path',
        'Dist-Zilla-Plugin-Foo' => 'share/DZPF',
    };

C<-root> is optional, and defaults to C<cwd>

B<NOTE:> There's a bug prior to 5.18 with C<< use Foo { -key => } >>, so for backwards compatibility, make sure you either quote
the key: C<< use Foo { '-key' => } >>, or make it the non-first key.

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"Test::File::ShareDir::Dist",
    "interface":"exporter"
}


=end MetaPOD::JSON

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Test/File/ShareDir/Utils.pm
use 5.006;
use strict;
use warnings;

package Test::File::ShareDir::Utils;

our $VERSION = '1.001003';

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

# ABSTRACT: Simple utilities for File::ShareDir testing

use Exporter 5.57 qw(import);
use Carp qw( croak );

our @EXPORT_OK = qw( extract_dashes );
























sub extract_dashes {
  my ( $undashed_to, $source ) = @_;
  if ( not ref $source or 'HASH' ne ref $source ) {
    return croak(q[Must pass a hashref]);
  }
  my %input_config = %{$source};
  my $params       = {};
  for my $key ( keys %input_config ) {
    next unless $key =~ /\A-(.*)\z/msx;
    $params->{$1} = delete $input_config{$key};
  }
  $params->{$undashed_to} = {} if not exists $params->{$undashed_to};
  for my $key ( keys %input_config ) {
    $params->{$undashed_to}->{$key} = $input_config{$key};
  }
  return $params;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::File::ShareDir::Utils - Simple utilities for File::ShareDir testing

=head1 VERSION

version 1.001003

=head1 SYNOPSIS

  use Test::File::ShareDir::Utils qw( extract_dashes );

  my $hash = extract_dashes('dists', $oldhash );

=head1 EXPORTABLE FUNCTIONS

=head2 extract_dashes

A utility that helps transform:

  -opt_a => bar
  -opt_b => baz
  NameA  => NameAValue
  NameB  => NameBValue

Into

  opt_a => bar
  opt_b => baz
  modules => {
    NameA => NameAValue
    NameB => NameBValue
  }

This is a useful approach used all over import and functional style interfaces due to explicit configuration
being needed only on rare occasions.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.012.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Create a Fake ShareDir for your modules for testing.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Test-File-ShareDir",
  "dist_version" => "0.4.2",
  "license" => "perl",
  "module_name" => "Test::File::ShareDir",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Class::Tiny" => 0,
    "File::Copy::Recursive" => 0,
    "File::ShareDir" => "1.00",
    "File::Temp" => 0,
    "Path::Tiny" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Cwd" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "FindBin" => 0,
    "List::Util" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.96",
    "lib" => 0,
    "version" => 0
  }
);


my %fallback_build_requires = (
  "Cwd" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec::Functions" => 0,
  "FindBin" => 0,
  "List::Util" => 0,
  "Module::Build" => "0.3601",
  "Test::Fatal" => 0,
  "Test::More" => "0.96",
  "lib" => 0,
  "version" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# FILENAME: bundle_to_ini.pl
# CREATED: 02/06/14 01:48:56 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Write an INI file from a bundle

use 5.008;    #utf8
use strict;
use warnings;
use utf8;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    ShareDir sharedirs dir dirs Notedly tempdir
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

$bundle->remove_policy('CodeLayout::RequireUseUTF8');
$bundle->remove_policy('NamingConventions::Capitalization');

#$bundle->remove_policy('Subroutines::ProhibitCallsToUnexportedSubs');
$bundle->remove_policy('Subroutines::ProhibitSubroutinePrototypes');

#$bundle->remove_policy('Subroutines::ProhibitExcessComplexity');

$bundle->remove_policy('Subroutines::RequireArgUnpacking');

#$bundle->remove_policy('ErrorHandling::RequireCarping');
$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  open my $rcfile, '>', './perlcritic.rc' or croak 'Cant open perlcritic.rc';
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  open my $depsfile, '>', './perlcritic.deps' or croak 'Cant open perlcritic.deps';
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.008.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Create a Fake ShareDir for your modules for testing.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Test-File-ShareDir",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Test::File::ShareDir",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Class::Tiny" => 0,
    "Exporter" => "5.57",
    "File::Copy::Recursive" => 0,
    "File::ShareDir" => "1.00",
    "Path::Tiny" => "0.018",
    "Scope::Guard" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Cwd" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.96",
    "lib" => 0
  },
  "VERSION" => "1.001003",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Class::Tiny" => 0,
  "Cwd" => 0,
  "Exporter" => "5.57",
  "ExtUtils::MakeMaker" => 0,
  "File::Copy::Recursive" => 0,
  "File::ShareDir" => "1.00",
  "File::Spec" => 0,
  "FindBin" => 0,
  "Path::Tiny" => "0.018",
  "Scope::Guard" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.96",
  "lib" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
